File: DailyLeetcode/April/April18.rs
use std::collections::HashSet;

impl Solution {
    pub fn island_perimeter(grid: Vec<Vec<i32>>) -> i32 {
        let mut result = 0;
        let mut visited: HashSet<(usize, usize)> = HashSet::new();
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if grid[i][j] == 1 {
                    result += dfs(&grid, i, j, &mut visited);
                }
            }
        }
        result
    }
}

// Define dfs outside the impl block
fn dfs(grid: &Vec<Vec<i32>>, x: usize, y: usize, visited: &mut HashSet<(usize, usize)>) -> i32 {
    if x >= grid.len() || y >= grid[0].len() || grid[x][y] == 0 {
        return 1;
    }
    let pos = (x, y);
    if visited.contains(&pos) {
        return 0;
    }

    visited.insert(pos);

    let mut result = 0;
    result += dfs(grid, x + 1, y, visited);
    result += dfs(grid, x - 1, y, visited);
    result += dfs(grid, x, y + 1, visited);
    result += dfs(grid, x, y - 1, visited);
    result
}

File: Main/DailyLeetcode/April/April18.cc
#include <set>
#include <vector>
using namespace std;

// slow but this is to practice DFS, also since we are using set instead of
// unordered_set
class Solution {
  set<pair<int, int>> visited;

public:
  int islandPerimeter(vector<vector<int>> &grid) {
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid[0].size(); j++) {
        if (grid[i][j] == 1) {
          return dfs(grid, i, j);
        }
      }
    }
    return 0;
  }

  int dfs(vector<vector<int>> &grid, int x, int y) {
    int result = 0;
    if (x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size() ||
        grid[x][y] == 0) {
      // we got here from a land square, so + 1
      return 1;
    }
    pair<int, int> position = {x, y};
    if (visited.find(position) != visited.end()) {
      return 0;
    }

    visited.insert(position);
    result += dfs(grid, x + 1, y);
    result += dfs(grid, x, y + 1);
    result += dfs(grid, x - 1, y);
    result += dfs(grid, x, y - 1);
    return result;
  }
};
